# 1. Import Necessary Library

In [2]:
import pandas as pd
from sklearn.metrics import pairwise_distances

import warnings 
warnings.filterwarnings('ignore')

# 2. Import Data

In [4]:
books_data = pd.read_csv('book.csv',encoding= 'unicode_escape')
books_data

Unnamed: 0  User.ID                                         Book.Title  \
0              1   276726                                Classical Mythology   
1              2   276729                                       Clara Callan   
2              3   276729                               Decision in Normandy   
3              4   276736  Flu: The Story of the Great Influenza Pandemic...   
4              5   276737                             The Mummies of Urumchi   
...          ...      ...                                                ...   
9995        9996   162121       American Fried: Adventures of a Happy Eater.   
9996        9997   162121                              Cannibal In Manhattan   
9997        9998   162121                    How to Flirt: A Practical Guide   
9998        9999   162121                                           Twilight   
9999       10000   162129                       Kids Say the Darndest Things   

      Book.Rating  
0               5  
1               3  
2               6  
3               8  
4               6  
...           ...  
9995            7  
9996            9  
9997            7  
9998            8  
9999            6  

[10000 rows x 4 columns]

# 3. Data Understanding

In [5]:
books_data.shape

(10000, 4)

In [6]:
books_data.dtypes

Unnamed: 0      int64
User.ID         int64
Book.Title     object
Book.Rating     int64
dtype: object

In [7]:
books_data.isna().sum()

Unnamed: 0     0
User.ID        0
Book.Title     0
Book.Rating    0
dtype: int64

In [8]:
books_data['User.ID'].nunique()

2182

In [9]:
books_data['User.ID'].unique()

array([276726, 276729, 276736, ..., 162113, 162121, 162129], dtype=int64)

In [10]:
books_data['Book.Title'].unique()

array(['Classical Mythology', 'Clara Callan', 'Decision in Normandy', ...,
       'How to Flirt: A Practical Guide', 'Twilight',
       'Kids Say the Darndest Things'], dtype=object)

In [11]:
books_data['Book.Title'].nunique()

9659

# 4. Data Preparation

In [14]:
user_pivot_table = pd.pivot_table(data = books_data,values = 'Book.Rating',index= 'User.ID',columns = 'Book.Title').fillna(0)
user_pivot_table.index = books_data['User.ID'].unique()
user_pivot_table

Book.Title   Jason, Madison &amp   Other Stories;Merril;1985;McClelland &amp  \
276726                       0.0                                         0.0   
276729                       0.0                                         0.0   
276736                       0.0                                         0.0   
276737                       0.0                                         0.0   
276744                       0.0                                         0.0   
...                          ...                                         ...   
162107                       0.0                                         0.0   
162109                       0.0                                         0.0   
162113                       0.0                                         0.0   
162121                       0.0                                         0.0   
162129                       0.0                                         0.0   

Book.Title   Repairing PC Drives &amp  '48  \
276726                            0.0  0.0   
276729                            0.0  0.0   
276736                            0.0  0.0   
276737                            0.0  0.0   
276744                            0.0  0.0   
...                               ...  ...   
162107                            0.0  0.0   
162109                            0.0  0.0   
162113                            0.0  0.0   
162121                            0.0  0.0   
162129                            0.0  0.0   

Book.Title  'O Au No Keia: Voices from Hawai'I's Mahu and Transgender Communities  \
276726                                                    0.0                       
276729                                                    0.0                       
276736                                                    0.0                       
276737                                                    0.0                       
276744                                                    0.0                       
...                                                       ...                       
162107                                                    0.0                       
162109                                                    0.0                       
162113                                                    0.0                       
162121                                                    0.0                       
162129                                                    0.0                       

Book.Title  ...AND THE HORSE HE RODE IN ON : THE PEOPLE V. KENNETH STARR  \
276726                                                    0.0              
276729                                                    0.0              
276736                                                    0.0              
276737                                                    0.0              
276744                                                    0.0              
...                                                       ...              
162107                                                    0.0              
162109                                                    0.0              
162113                                                    0.0              
162121                                                    0.0              
162129                                                    0.0              

Book.Title  01-01-00: A Novel of the Millennium  \
276726                                      0.0   
276729                                      0.0   
276736                                      0.0   
276737                                      0.0   
276744                                      0.0   
...                                         ...   
162107                                      0.0   
162109                                      0.0   
162113                                      0.0   
162121                                    

# 5. Model Building

## 5.1 Cosine Matric

In [17]:
cosine_metrices = 1 - pairwise_distances(X = user_pivot_table, metric='cosine' ).round(3)
cosine_metrices_df = pd.DataFrame(cosine_metrices)
cosine_metrices_df.index = books_data['User.ID'].unique()
cosine_metrices_df.columns = books_data['User.ID'].unique()
cosine_metrices_df.head(25)

276726  276729  276736  276737  276744  276745  276747  276748  \
276726     1.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
276729     0.0     1.0     0.0     0.0     0.0     0.0     0.0     0.0   
276736     0.0     0.0     1.0     0.0     0.0     0.0     0.0     0.0   
276737     0.0     0.0     0.0     1.0     0.0     0.0     0.0     0.0   
276744     0.0     0.0     0.0     0.0     1.0     0.0     0.0     0.0   
276745     0.0     0.0     0.0     0.0     0.0     1.0     0.0     0.0   
276747     0.0     0.0     0.0     0.0     0.0     0.0     1.0     0.0   
276748     0.0     0.0     0.0     0.0     0.0     0.0     0.0     1.0   
276751     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
276754     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
276755     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
276760     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
276762     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
276768     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
276772     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
276774     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
276780     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
276786     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
276788     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
276796     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
276798     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
276800     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
276804     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
276808     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
276811     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   

        276751  276754  ...  162085  162091  162092  162095  162103  162107  \
276726     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276729     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276736     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276737     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276744     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276745     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276747     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276748     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276751     1.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276754     0.0     1.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276755     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276760     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276762     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276768     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276772     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276774     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276780     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276786     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276788     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276796     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276798     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276800     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276804     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276808     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276811     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   

        162109  162113  1621

## From the cosine similarity recommendation model we can conclude there can be no recommendation done on the basis of similarity as the books are not similar in nature as far as we can see.

## For a better Recommendation system we can try different models as well.

## 5.2 Euclideon Distance

In [19]:
euclideon_distance_user = pairwise_distances(X = user_pivot_table,metric='euclidean').round()
euclideon_distance_user_df = pd.DataFrame(data = euclideon_distance_user)
euclideon_distance_user_df.columns = books_data['User.ID'].unique()
euclideon_distance_user_df.index   = books_data['User.ID'].unique()
euclideon_distance_user_df

276726  276729  276736  276737  276744  276745  276747  276748  \
276726     0.0    16.0    16.0    18.0    18.0    17.0    18.0    16.0   
276729    16.0     0.0     8.0    12.0    11.0    11.0    12.0     9.0   
276736    16.0     8.0     0.0    12.0    11.0    11.0    12.0     9.0   
276737    18.0    12.0    12.0     0.0    14.0    13.0    15.0    12.0   
276744    18.0    11.0    11.0    14.0     0.0    13.0    14.0    12.0   
...        ...     ...     ...     ...     ...     ...     ...     ...   
162107    17.0    10.0    10.0    13.0    12.0    12.0    14.0    11.0   
162109    17.0    11.0    11.0    13.0    13.0    13.0    14.0    11.0   
162113    29.0    26.0    26.0    27.0    27.0    27.0    27.0    26.0   
162121    17.0    10.0    10.0    13.0    12.0    12.0    14.0    11.0   
162129    23.0    18.0    18.0    20.0    20.0    19.0    20.0    19.0   

        276751  276754  ...  162085  162091  162092  162095  162103  162107  \
276726    16.0    20.0  ...    18.0    23.0    17.0    38.0    18.0    17.0   
276729     9.0    15.0  ...    12.0    18.0    10.0    35.0    13.0    10.0   
276736     9.0    15.0  ...    12.0    18.0    10.0    35.0    13.0    10.0   
276737    12.0    17.0  ...    14.0    20.0    13.0    36.0    15.0    13.0   
276744    12.0    16.0  ...    14.0    20.0    12.0    36.0    14.0    12.0   
...        ...     ...  ...     ...     ...     ...     ...     ...     ...   
162107    11.0    16.0  ...    13.0    19.0    11.0    36.0    14.0     0.0   
162109    11.0    16.0  ...    13.0    20.0    12.0    36.0    14.0    12.0   
162113    26.0    28.0  ...    27.0    30.0    26.0    43.0    27.0    26.0   
162121    11.0    16.0  ...    13.0    19.0    11.0    36.0    14.0    11.0   
162129    19.0    22.0  ...    20.0    24.0    19.0    39.0    20.0    19.0   

        162109  162113  162121  162129  
276726    17.0    29.0    17.0    23.0  
276729    11.0    26.0    10.0    18.0  
276736    11.0    26.0    10.0    18.0  
276737    13.0    27.0    13.0    20.0  
276744    13.0    27.0    12.0    20.0  
...        ...     ...     ...     ...  
162107    12.0    26.0    11.0    19.0  
162109     0.0    27.0    12.0    19.0  
162113    27.0     0.0    26.0    30.0  
162121    12.0    26.0     0.0    19.0  
162129    19.0    30.0    19.0     0.0  

[2182 rows x 2182 columns]

## In this model we look into the distance between two features or books, lesser the distance more correlated the items are and more chances for recommendation. For eg in the above model book number 276751 and 276729 have distance of 9 therefore they can be recommended to the user whereas book 276726 and 276736 have a distance of 16 therefore it cannot be recommended to the user.

## 5.3 correlation metrices

In [20]:
user_pivot_table_2 = pd.pivot_table(data = books_data,values = 'Book.Rating',index= 'Book.Title',columns = 'User.ID').fillna(0)
user_pivot_table_2.columns = books_data['User.ID'].unique()
user_pivot_table_2

276726  276729  276736  \
Book.Title                                                                   
 Jason, Madison &amp                                   0.0     0.0     0.0   
 Other Stories;Merril;1985;McClelland &amp             0.0     0.0     0.0   
 Repairing PC Drives &amp                              0.0     0.0     0.0   
'48                                                    0.0     0.0     0.0   
'O Au No Keia: Voices from Hawai'I's Mahu and T...     0.0     0.0     0.0   
...                                                    ...     ...     ...   
\Surely You're Joking, Mr. Feynman!\: Adventure...     0.0     0.0     0.0   
\Well, there's your problem\: Cartoons                 0.0     0.0     0.0   
iI Paradiso Degli Orchi                                0.0     0.0     0.0   
stardust                                               0.0     0.0     0.0   
Ã?Â?bermorgen.                                         0.0     0.0     0.0   

                                                    276737  276744  276745  \
Book.Title                                                                   
 Jason, Madison &amp                                   0.0     0.0     0.0   
 Other Stories;Merril;1985;McClelland &amp             0.0     0.0     0.0   
 Repairing PC Drives &amp                              0.0     0.0     0.0   
'48                                                    0.0     0.0     0.0   
'O Au No Keia: Voices from Hawai'I's Mahu and T...     0.0     0.0     0.0   
...                                                    ...     ...     ...   
\Surely You're Joking, Mr. Feynman!\: Adventure...     0.0     0.0     0.0   
\Well, there's your problem\: Cartoons                 0.0     0.0     0.0   
iI Paradiso Degli Orchi                                0.0     0.0     0.0   
stardust                                               0.0     0.0     0.0   
Ã?Â?bermorgen.                                         0.0     0.0     0.0   

                                                    276747  276748  276751  \
Book.Title                                                                   
 Jason, Madison &amp                                   0.0     0.0     0.0   
 Other Stories;Merril;1985;McClelland &amp             0.0     0.0     0.0   
 Repairing PC Drives &amp                              0.0     0.0     0.0   
'48                                                    0.0     0.0     0.0   
'O Au No Keia: Voices from Hawai'I's Mahu and T...     0.0     0.0     0.0   
...                                                    ...     ...     ...   
\Surely You're Joking, Mr. Feynman!\: Adventure...     0.0     0.0     0.0   
\Well, there's your problem\: Cartoons                 0.0     0.0     0.0   
iI Paradiso Degli Orchi                                0.0     0.0     0.0   
stardust                                               0.0     0.0     0.0   
Ã?Â?bermorgen.                                         0.0     0.0     0.0   

                                                    276754  ...  162085  \
Book.Title                                                  ...           
 Jason, Madison &amp                                   0.0  ...     0.0   
 Other Stories;Merril;1985;McClelland &amp             0.0  ...     0.0   
 Repairing PC Drives &amp                              0.0  ...     0.0   
'48                                                    0.0  ...     0.0   
'O Au No Keia: Voices from Hawai'I's Mahu and T...     0.0  ...     0.0   
...                                                    ...  ...     ...   
\Surely You're Joking, Mr. Feynman!\: Adventure...     0.0  ...     0.0   
\Well, there's your problem\: Cartoons                 0.0  ...     0.0   
iI Paradiso Degli Orchi                                0.0  ...     0.0   
stardust                                               0.0  ...     0.0   
Ã?Â?bermorgen.                                         0.0  ...     0.0   

                               

In [21]:
user_pivot_table_2.corr().round(2)

276726  276729  276736  276737  276744  276745  276747  276748  \
276726     1.0    -0.0    -0.0    -0.0    -0.0    -0.0    -0.0    -0.0   
276729    -0.0     1.0    -0.0    -0.0    -0.0    -0.0    -0.0    -0.0   
276736    -0.0    -0.0     1.0    -0.0    -0.0    -0.0    -0.0    -0.0   
276737    -0.0    -0.0    -0.0     1.0    -0.0    -0.0    -0.0    -0.0   
276744    -0.0    -0.0    -0.0    -0.0     1.0    -0.0    -0.0    -0.0   
...        ...     ...     ...     ...     ...     ...     ...     ...   
162107    -0.0    -0.0    -0.0    -0.0    -0.0    -0.0    -0.0    -0.0   
162109    -0.0    -0.0    -0.0    -0.0    -0.0    -0.0    -0.0    -0.0   
162113    -0.0    -0.0    -0.0    -0.0    -0.0    -0.0    -0.0    -0.0   
162121    -0.0    -0.0    -0.0    -0.0    -0.0    -0.0    -0.0    -0.0   
162129    -0.0    -0.0    -0.0    -0.0    -0.0    -0.0    -0.0    -0.0   

        276751  276754  ...  162085  162091  162092  162095  162103  162107  \
276726    -0.0    -0.0  ...    -0.0    -0.0    -0.0    -0.0    -0.0    -0.0   
276729    -0.0    -0.0  ...    -0.0    -0.0    -0.0    -0.0    -0.0    -0.0   
276736    -0.0    -0.0  ...    -0.0    -0.0    -0.0    -0.0    -0.0    -0.0   
276737    -0.0    -0.0  ...    -0.0    -0.0    -0.0    -0.0    -0.0    -0.0   
276744    -0.0    -0.0  ...    -0.0    -0.0    -0.0    -0.0    -0.0    -0.0   
...        ...     ...  ...     ...     ...     ...     ...     ...     ...   
162107    -0.0    -0.0  ...    -0.0    -0.0    -0.0    -0.0    -0.0     1.0   
162109    -0.0    -0.0  ...    -0.0    -0.0    -0.0    -0.0    -0.0    -0.0   
162113    -0.0    -0.0  ...    -0.0    -0.0    -0.0    -0.0    -0.0    -0.0   
162121    -0.0    -0.0  ...    -0.0    -0.0    -0.0    -0.0    -0.0    -0.0   
162129    -0.0    -0.0  ...    -0.0    -0.0    -0.0    -0.0    -0.0    -0.0   

        162109  162113  162121  162129  
276726    -0.0    -0.0    -0.0    -0.0  
276729    -0.0    -0.0    -0.0    -0.0  
276736    -0.0    -0.0    -0.0    -0.0  
276737    -0.0    -0.0    -0.0    -0.0  
276744    -0.0    -0.0    -0.0    -0.0  
...        ...     ...     ...     ...  
162107    -0.0    -0.0    -0.0    -0.0  
162109     1.0    -0.0    -0.0    -0.0  
162113    -0.0     1.0    -0.0    -0.0  
162121    -0.0    -0.0     1.0    -0.0  
162129    -0.0    -0.0    -0.0     1.0  

[2182 rows x 2182 columns]

In [25]:
books_data[books_data['User.ID'] == 276726]

Unnamed: 0  User.ID           Book.Title  Book.Rating
0           1   276726  Classical Mythology            5

In [26]:
books_data[books_data['User.ID'] == 276729]

Unnamed: 0  User.ID            Book.Title  Book.Rating
1           2   276729          Clara Callan            3
2           3   276729  Decision in Normandy            6

# ======================== THE END =========================